File: folio/builtins.py
"""Single source of truth for Folio's builtin MDX components.

These components ship inside the bundled Next/Nextra template
(``template/components/``) and are registered into the
:class:`~folio.extensions.ExtensionRegistry` so that the registry is the
authoritative description of every component Folio exposes. The MDX contract
(:mod:`folio.generator.mdx_contract`) and the build-time component injection
both derive from this manifest.

The order of :data:`BUILTIN_COMPONENTS` MUST match the entry order in
``template/mdx-components.tsx``; ``tests/test_builtin_drift.py`` guards this.

Contract membership is explicit: entries with ``contract=True`` are part of
the published MDX component contract, and ``source_label`` carries the
contract ``source`` string for those entries. ``category`` is pure taxonomy
and plays no role in the contract. The prop type strings are copied verbatim
from the hand-authored contract.
"""

from __future__ import annotations

from folio.extensions import ComponentDefinition, ExtensionRegistry


def _component(
    name: str,
    module: str,
    *,
    props: dict[str, str] | None = None,
    required: bool = False,
    category: str = "component-catalog",
    contract: bool = False,
    source_label: str = "",
) -> ComponentDefinition:
    return ComponentDefinition(
        name=name,
        import_path=f"@/components/{module}",
        expose_mdx=True,
        props=props or {},
        required=required,
        category=category,
        contract=contract,
        source_label=source_label,
        origin="builtin",
    )


BUILTIN_COMPONENTS: tuple[ComponentDefinition, ...] = (
    _component(
        "ParamTable",
        "param-table",
        required=True,
        category="api-reference",
        contract=True,
        source_label="api-reference",
        props={
            "args": (
                "Array<{ name: string; type: string; default?: string; "
                "description?: string | null; href?: string }>"
            )
        },
    ),
    _component(
        "ClassOverview",
        "class-overview",
        required=True,
        category="api-reference",
        contract=True,
        source_label="api-reference",
        props={
            "name": "string",
            "bases": "string[] | Array<{ name: string; href?: string }>",
            "decorators": "string[]",
            "description": "string",
        },
    ),
    _component(
        "TypeBadge",
        "type-badge",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"type": "string", "href": "string"},
    ),
    _component(
        "MethodAccordion",
        "method-accordion",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={
            "methods": (
                "Array<{ name: string; signature: string; description?: string }>"
            )
        },
    ),
    _component("ExampleTabs", "example-tabs"),
    _component("DeprecationNotice", "deprecation-notice"),
    _component(
        "Callout",
        "callout",
        required=True,
        category="markdown-rst",
        contract=True,
        source_label="markdown-rst",
        props={
            "type": '"note" | "warning" | "info" | "tip" | "check" | "danger"',
            "title": "string",
            "children": "React.ReactNode",
        },
    ),
    _component("CodeGroup", "code-group"),
    _component(
        "SourceLink",
        "source-link",
        required=True,
        category="api-reference",
        contract=True,
        source_label="api-reference",
        props={"href": "string"},
    ),
    _component("Steps", "steps"),
    _component("Step", "steps"),
    _component(
        "Mermaid",
        "mermaid",
        required=True,
        category="markdown-mdx",
        contract=True,
        source_label="markdown-mdx",
        props={"chart": "string"},
    ),
    _component(
        "FileTree",
        "file-tree",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"tree": "string"},
    ),
    _component(
        "FeatureCard",
        "feature-card",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"title": "string", "description": "string", "href": "string"},
    ),
    _component(
        "CardGrid",
        "card-grid",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"columns": "number", "children": "React.ReactNode"},
    ),
    _component(
        "Tabs",
        "tabs",
        required=True,
        category="markdown-mdx",
        contract=True,
        source_label="markdown-mdx",
        props={"defaultValue": "string", "children": "React.ReactNode"},
    ),
    _component(
        "TabItem",
        "tabs",
        required=True,
        category="markdown-mdx",
        contract=True,
        source_label="markdown-mdx",
        props={
            "label": "string",
            "value": "string",
            "children": "React.ReactNode",
        },
    ),
    _component(
        "Accordion",
        "accordion",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"children": "React.ReactNode"},
    ),
    _component(
        "AccordionItem",
        "accordion",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"title": "string", "children": "React.ReactNode"},
    ),
    _component(
        "Timeline",
        "timeline",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={"children": "React.ReactNode"},
    ),
    _component(
        "TimelineItem",
        "timeline",
        category="component-catalog",
        contract=True,
        source_label="component-catalog",
        props={
            "date": "string",
            "title": "string",
            "badge": "string",
            "description": "string",
        },
    ),
    _component("TerminalSession", "terminal-session"),
    _component("ConfigPanel", "config-panel"),
    _component("BuildArtifact", "build-artifact"),
    _component("DocPreview", "doc-preview"),
    _component("PreviewCode", "preview-code"),
    _component("CommandGrid", "command-grid"),
    _component("CommandCard", "command-grid"),
    _component("BeforeAfter", "before-after"),
    _component(
        "Swot",
        "swot",
        contract=True,
        source_label="component-catalog",
        props={
            "strengths": "string[]",
            "weaknesses": "string[]",
            "opportunities": "string[]",
            "threats": "string[]",
            "title": "string | undefined",
        },
    ),
    _component(
        "CompareMatrix",
        "compare-matrix",
        contract=True,
        source_label="component-catalog",
        props={
            "tools": "string[]",
            "rows": "{ feature: string; values: (boolean | string)[]; note?: string }[]",
            "caption": "string | undefined",
            "highlight": "number | undefined",
        },
    ),
    _component(
        "PullQuote",
        "pull-quote",
        contract=True,
        source_label="component-catalog",
        props={
            "children": "ReactNode",
            "kicker": "string | undefined",
            "attribution": "string | undefined",
        },
    ),
    _component(
        "StatStrip",
        "stat-strip",
        contract=True,
        source_label="component-catalog",
        props={
            "stats": "{ value: string; label: string; detail?: string }[]",
        },
    ),
    _component("Checklist", "checklist"),
    _component("HookMap", "hook-map"),
    _component("ComponentIndex", "component-index"),
    _component(
        "ApiReferenceIndex",
        "api-reference-index",
        required=True,
        category="api-reference",
        contract=True,
        source_label="api-reference",
        props={
            "modules": (
                "Array<{ name: string; classes: number; "
                "functions: number; path: string }>"
            )
        },
    ),
    _component("ComparisonMatrix", "comparison-matrix"),
    _component("UnavailableFeature", "unavailable-feature"),
    _component(
        "BrowserFrame",
        "browser-frame",
        contract=True,
        source_label="component-catalog",
        props={
            "url": "string",
            "label": "string | undefined",
            "children": "React.ReactNode",
        },
    ),
)


def register_builtin_components(registry: ExtensionRegistry) -> None:
    """Register every builtin component into ``registry`` in manifest order."""
    for component in BUILTIN_COMPONENTS:
        registry.register_component(
            component.name,
            import_path=component.import_path,
            export_name=component.export_name,
            expose_mdx=component.expose_mdx,
            props=component.props,
            required=component.required,
            category=component.category,
            contract=component.contract,
            source_label=component.source_label,
            origin="builtin",
        )

File: crates/folio-plugins/src/builtins.rs
//! The builtin component manifest, the public layout and their registration.
//! The manifest order is the entry order of `template/mdx-components.tsx`;
//! `drift.rs` guards it.

use std::sync::LazyLock;

use indexmap::IndexMap;

use crate::error::PluginsError;
use crate::registry::{ComponentDefinition, ComponentOrigin, ExtensionRegistry};

/// The one layout every public view renders inside (internal name, never emitted).
pub const PUBLIC_LAYOUT: &str = "folio.public";

/// A manifest entry: `contract` names the source label of a contract member
/// (its category too); `None` is a catalog-only component with no props.
fn component(
    name: &str,
    module: &str,
    required: bool,
    contract: Option<&str>,
    props: &[(&str, &str)],
) -> ComponentDefinition {
    ComponentDefinition {
        required,
        category: contract.unwrap_or("component-catalog").to_string(),
        contract: contract.is_some(),
        source_label: contract.unwrap_or("").to_string(),
        props: props
            .iter()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect::<IndexMap<_, _>>(),
        origin: ComponentOrigin::Builtin,
        ..ComponentDefinition::new(name, &format!("@/components/{module}"))
    }
}

const API: Option<&str> = Some("api-reference");
const RST: Option<&str> = Some("markdown-rst");
const MDX: Option<&str> = Some("markdown-mdx");
const CATALOG: Option<&str> = Some("component-catalog");

/// The forty builtin components in `mdx-components.tsx` entry order.
pub static BUILTIN_COMPONENTS: LazyLock<Vec<ComponentDefinition>> = LazyLock::new(|| {
    vec![
        component("ParamTable", "param-table", true, API, &[(
            "args",
            "Array<{ name: string; type: string; default?: string; description?: string | null; href?: string }>",
        )]),
        component("ClassOverview", "class-overview", true, API, &[
            ("name", "string"),
            ("bases", "string[] | Array<{ name: string; href?: string }> | undefined"),
            ("decorators", "string[] | undefined"),
            ("description", "string | undefined"),
        ]),
        component("TypeBadge", "type-badge", false, CATALOG, &[
            ("type", "string"),
            ("href", "string | undefined"),
        ]),
        component("MethodAccordion", "method-accordion", false, CATALOG, &[(
            "methods",
            "Array<{ name: string; signature: string; description: string; isAsync?: boolean; children?: React.ReactNode }>",
        )]),
        component("ExampleTabs", "example-tabs", false, None, &[]),
        component("DeprecationNotice", "deprecation-notice", false, None, &[]),
        component("Callout", "callout", true, RST, &[
            ("type", "\"note\" | \"warning\" | \"info\" | \"tip\" | \"check\" | \"danger\" | undefined"),
            ("title", "string | undefined"),
            ("children", "React.ReactNode"),
        ]),
        component("CodeGroup", "code-group", false, None, &[]),
        component("SourceLink", "source-link", true, API, &[("href", "string")]),
        component("Steps", "steps", false, None, &[]),
        component("Step", "steps", false, None, &[]),
        component("Mermaid", "mermaid", true, MDX, &[("chart", "string")]),
        component("FileTree", "file-tree", false, CATALOG, &[("tree", "string")]),
        component("FeatureCard", "feature-card", false, CATALOG, &[
            ("title", "string"),
            ("description", "string"),
            ("icon", "string | undefined"),
            ("href", "string | undefined"),
        ]),
        component("CardGrid", "card-grid", false, CATALOG, &[
            ("columns", "2 | 3 | 4 | undefined"),
            ("children", "React.ReactNode"),
        ]),
        component("Tabs", "tabs", true, MDX, &[
            ("children", "React.ReactNode"),
            ("aria-label", "string | undefined"),
        ]),
        component("TabItem", "tabs", true, MDX, &[
            ("label", "string"),
            ("children", "React.ReactNode"),
        ]),
        component("Accordion", "accordion", false, CATALOG, &[("children", "React.ReactNode")]),
        component("AccordionItem", "accordion", false, CATALOG, &[
            ("title", "string"),
            ("children", "React.ReactNode"),
            ("defaultOpen", "boolean | undefined"),
        ]),
        component("Timeline", "timeline", false, CATALOG, &[("children", "React.ReactNode")]),
        component("TimelineItem", "timeline", false, CATALOG, &[
            ("date", "string"),
            ("title", "string"),
            ("badge", "string | undefined"),
            ("children", "React.ReactNode"),
        ]),
        component("TerminalSession", "terminal-session", false, None, &[]),
        component("ConfigPanel", "config-panel", false, None, &[]),
        component("BuildArtifact", "build-artifact", false, None, &[]),
        component("DocPreview", "doc-preview", false, None, &[]),
        component("PreviewCode", "preview-code", false, None, &[]),
        component("CommandGrid", "command-grid", false, None, &[]),
        component("CommandCard", "command-grid", false, None, &[]),
        component("BeforeAfter", "before-after", false, None, &[]),
        component("Swot", "swot", false, CATALOG, &[
            ("strengths", "string[]"),
            ("weaknesses", "string[]"),
            ("opportunities", "string[]"),
            ("threats", "string[]"),
            ("title", "string | undefined"),
        ]),
        component("CompareMatrix", "compare-matrix", false, CATALOG, &[
            ("tools", "string[]"),
            ("rows", "{ feature: string; values: (boolean | string)[]; note?: string }[]"),
            ("caption", "string | undefined"),
            ("highlight", "number | undefined"),
        ]),
        component("PullQuote", "pull-quote", false, CATALOG, &[
            ("children", "ReactNode"),
            ("kicker", "string | undefined"),
            ("attribution", "string | undefined"),
        ]),
        component("StatStrip", "stat-strip", false, CATALOG, &[(
            "stats",
            "{ value: string; label: string; detail?: string }[]",
        )]),
        component("Checklist", "checklist", false, None, &[]),
        component("HookMap", "hook-map", false, None, &[]),
        component("ComponentIndex", "component-index", false, None, &[]),
        component("ApiReferenceIndex", "api-reference-index", true, API, &[(
            "modules",
            "Array<{ name: string; description: string; href: string; classCount: number; functionCount: number }>",
        )]),
        component("ComparisonMatrix", "comparison-matrix", false, None, &[]),
        component("UnavailableFeature", "unavailable-feature", false, None, &[]),
        component("BrowserFrame", "browser-frame", false, CATALOG, &[
            ("url", "string"),
            ("label", "string | undefined"),
            ("footer", "React.ReactNode | undefined"),
            ("children", "React.ReactNode | undefined"),
        ]),
    ]
});

/// Register every builtin component in manifest order.
pub fn register_builtin_components(registry: &mut ExtensionRegistry) -> Result<(), PluginsError> {
    for component in BUILTIN_COMPONENTS.iter() {
        registry.register_component(component.clone(), &mut Vec::new())?;
    }
    Ok(())
}

/// Register the `PUBLIC_LAYOUT` layout unless it is already present.
pub fn register_builtin_extensions(registry: &mut ExtensionRegistry) -> Result<(), PluginsError> {
    if !registry.layouts.contains_key(PUBLIC_LAYOUT) {
        registry.register_layout(
            PUBLIC_LAYOUT,
            "@/components/folio-view-layouts",
            "PublicLayout",
            &["main"],
        )?;
    }
    Ok(())
}

#[cfg(test)]
#[path = "builtins_tests.rs"]
mod tests;
